Crash report: An error condition was reached, which should be statically unreachable. Assigning to a subscript sho

# Crash Message

Hello, I got the error message below with the following input program which is a slight modification of the array-clearing-succeeding example:

```
// -*- tab-width:2 ; indent-tabs-mode:nil -*-
//:: cases ZeroArrayIC
//:: tools silicon
//:: verdict Pass

public class ZeroArray {
    /*@ 
        context_everywhere ar != null;
        requires Perm(ar[*],1);
        ensures Perm(ar[*],1) ** (\forall int k ; 0 <= k && k < ar.length ; ar[k]==0 ) ;
     */
    public void zero_array(int ar[]){
        int N=ar.length;
        for(int i=0;i<N;i++)
        /*@  requires Perm(ar[i],1);
             ensures  Perm(ar[i],1) ** ar[i]==0;
        @*/
		        {
            ar[i*2]=0;
        }
		for(int i=1;i<N;i++)
        /*@  requires Perm(ar[i],1);
             ensures  Perm(ar[i],1) ** ar[i]==0;
        @*/

        {
            ar[i*2]=0;
        }
    }
}

```

on the VersCor website

https://vercors.ewi.utwente.nl/

```
An error condition was reached, which should be statically unreachable. Assigning to a subscript should trigger an error on the assignment, and not on the subscript.. Inner failure:
 > ======================================
 > At test.java
 > --------------------------------------
 >    25  
 >    26          {
 >                   [-------
 >    27              ar[i*2]=0;
 >                    -------]
 >    28          }
 >    29      }
 > --------------------------------------
 > Index may be negative, or exceed the length of the array. (https://utwente.nl/vercors#arrayBounds)
 > ======================================
  at vct.col.origin.PanicBlame.blame(Blame.scala:1460)
  at vct.col.rewrite.adt.ImportArray$ArrayBoundsPreconditionFailed.blame(ImportArray.scala:27)
  at vct.col.rewrite.adt.ImportArray$ArrayBoundsPreconditionFailed.blame(ImportArray.scala:22)
  at vct.col.origin.NoContext.blame(Blame.scala:1592)
  at vct.col.origin.NoContext.blame(Blame.scala:1588)
  at viper.api.backend.SilverBackend.processError(SilverBackend.scala:207)
  at viper.api.backend.SilverBackend.processError$(SilverBackend.scala:172)
  at viper.api.backend.silicon.Silicon.processError(Silicon.scala:34)
  at viper.api.backend.SilverBackend.$anonfun$submit$1(SilverBackend.scala:164)
  at viper.api.backend.SilverBackend.$anonfun$submit$1$adapted(SilverBackend.scala:164)
  at scala.collection.immutable.List.foreach(List.scala:333)
  at viper.api.backend.SilverBackend.submit(SilverBackend.scala:164)
  ...
  at vct.main.Main$.main(Main.scala:50)
  at vct.main.Main.main(Main.scala)
```

# Version Information
* VerCors version `2.3.0`
* At commit 4bd0a112c13d35dc64e9b357e063afa5af678e6b from branch `HEAD` (changes=false)

# Arguments
`/tmp/2aeef290-57e7-11f0-8239-9fbcc99b708f/test.java` `--progress`

# File Inputs
<details>
<summary>test.java</summary>

```java
// -*- tab-width:2 ; indent-tabs-mode:nil -*-
//:: cases ZeroArrayIC
//:: tools silicon
//:: verdict Pass

public class ZeroArray {
    /*@ 
        context_everywhere ar != null;
        requires Perm(ar[*],1);
        ensures Perm(ar[*],1) ** (\forall int k ; 0 <= k && k < ar.length ; ar[k]==0 ) ;
     */
    public void zero_array(int ar[]){
        int N=ar.length;
        for(int i=0;i<N;i++)
        /*@  requires Perm(ar[i],1);
             ensures  Perm(ar[i],1) ** ar[i]==0;
        @*/
		        {
            ar[i*2]=0;
        }
		for(int i=1;i<N;i++)
        /*@  requires Perm(ar[i],1);
             ensures  Perm(ar[i],1) ** ar[i]==0;
        @*/

        {
            ar[i*2]=0;
        }
    }
}


```
</details>


# Full Log
<details>

```
08:24:43.190 [INFO] Start: VerCors (at 08:24:43)
08:25:30.021 [INFO] Done: VerCors (at 08:25:30, duration: 00:00:46)
08:25:30.078 [WARN] Non-daemon thread UnknownThreadGroup.[VerCors] Block layout updates (#14) is still running
08:25:30.079 [WARN] Due to this VerCors will not exit and sit idly until that thread is done. You may want to stop the program manually.
08:25:30.099 [ERROR] vct.col.origin.BlameUnreachable: An error condition was reached, which should be statically unreachable. Assigning to a subscript should trigger an error on the assignment, and not on the subscript.. Inner failure:
 > ======================================
 > At test.java
 > --------------------------------------
 >    25  
 >    26          {
 >                   [-------
 >    27              ar[i*2]=0;
 >                    -------]
 >    28          }
 >    29      }
 > --------------------------------------
 > Index may be negative, or exceed the length of the array. (https://utwente.nl/vercors#arrayBounds)
 > ======================================
	at vct.col.origin.PanicBlame.blame(Blame.scala:1460)
	at vct.col.rewrite.adt.ImportArray$ArrayBoundsPreconditionFailed.blame(ImportArray.scala:27)
	at vct.col.rewrite.adt.ImportArray$ArrayBoundsPreconditionFailed.blame(ImportArray.scala:22)
	at vct.col.origin.NoContext.blame(Blame.scala:1592)
	at vct.col.origin.NoContext.blame(Blame.scala:1588)
	at viper.api.backend.SilverBackend.processError(SilverBackend.scala:207)
	at viper.api.backend.SilverBackend.processError$(SilverBackend.scala:172)
	at viper.api.backend.silicon.Silicon.processError(Silicon.scala:34)
	at viper.api.backend.SilverBackend.$anonfun$submit$1(SilverBackend.scala:164)
	at viper.api.backend.SilverBackend.$anonfun$submit$1$adapted(SilverBackend.scala:164)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at viper.api.backend.SilverBackend.submit(SilverBackend.scala:164)
	at viper.api.backend.SilverBackend.submit$(SilverBackend.scala:143)
	at viper.api.backend.silicon.Silicon.submit(Silicon.scala:34)
	at vct.main.stages.SilverBackend.verify(Backend.scala:201)
	at vct.main.stages.Backend.$anonfun$run$6(Backend.scala:166)
	at vct.main.stages.Backend.$anonfun$cachedDefinitelyVerifiesOrElseUpdate$1(Backend.scala:106)
	at scala.Option.getOrElse(Option.scala:201)
	at vct.main.stages.Backend.cachedDefinitelyVerifiesOrElseUpdate(Backend.scala:104)
	at vct.main.stages.Backend.cachedDefinitelyVerifiesOrElseUpdate$(Backend.scala:100)
	at vct.main.stages.SilverBackend.cachedDefinitelyVerifiesOrElseUpdate(Backend.scala:184)
	at vct.main.stages.Backend.$anonfun$run$5(Backend.scala:166)
	at vct.main.stages.Backend.$anonfun$run$5$adapted(Backend.scala:163)
	at hre.progress.Progress$.$anonfun$foreach$2(Progress.scala:25)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at hre.progress.task.AbstractTask.frame(AbstractTask.scala:166)
	at hre.progress.Progress$.$anonfun$foreach$1(Progress.scala:25)
	at hre.progress.Progress$.$anonfun$foreach$1$adapted(Progress.scala:24)
	at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:576)
	at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:574)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1300)
	at hre.progress.Progress$.foreach(Progress.scala:24)
	at vct.main.stages.Backend.$anonfun$run$1(Backend.scala:163)
	at hre.progress.task.NameSequenceTask.scope(NameSequenceTask.scala:16)
	at hre.progress.Progress$.stages(Progress.scala:47)
```

As the website encouraged me to report this here
```
[ERROR] !*!*!*!*!*!*!*!*!*!*!*!
[ERROR] ! VerCors has crashed !
[ERROR] !*!*!*!*!*!*!*!*!*!*!*!
[ERROR] Please report this as a bug here:
[ERROR] https://github.com/utwente-fmt/vercors/issues/new?labels=A-Bug&title=Crash+report%3A+An+error+condition+was+reached%2C+which+should+be+statically+unreachable.+Assigning+to+a+subscript+sho&body=%23+Crash+Message%0A%60%60%60%0AAn+error+condition+was+reached%2C+which+should+be+statically+unreachable.+Assigning+to+a+subscript+should+trigger+an+error+on+the+assignment%2C+and+not+on+the+subscript..+Inner+failure%3A%0A+%3E+%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%0A+%3E+At+test.java%0A+%3E+--------------------------------------%0A+%3E++++25++%0A+%3E++++26++++++++++%7B%0A+%3E+++++++++++++++++++%5B-------%0A+%3E++++27++++++++++++++ar%5Bi*2%5D%3D0%3B%0A+%3E++++++++++++++++++++-------%5D%0A+%3E++++28++++++++++%7D%0A+%3E++++29++++++%7D%0A+%3E+--------------------------------------%0A+%3E+Index+may+be+negative%2C+or+exceed+the+length+of+the+array.+%28https%3A%2F%2Futwente.nl%2Fvercors%23arrayBounds%29%0A+%3E+%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%0A++at+vct.col.origin.PanicBlame.blame%28Blame.scala%3A1460%29%0A++at+vct.col.rewrite.adt.ImportArray%24ArrayBoundsPreconditionFailed.blame%28ImportArray.scala%3A27%29%0A++at+vct.col.rewrite.adt.ImportArray%24ArrayBoundsPreconditionFailed.blame%28ImportArray.scala%3A22%29%0A++at+vct.col.origin.NoContext.blame%28Blame.scala%3A1592%29%0A++at+vct.col.origin.NoContext.blame%28Blame.scala%3A1588%29%0A++at+viper.api.backend.SilverBackend.processError%28SilverBackend.scala%3A207%29%0A++at+viper.api.backend.SilverBackend.processError%24%28SilverBackend.scala%3A172%29%0A++at+viper.api.backend.silicon.Silicon.processError%28Silicon.scala%3A34%29%0A++at+viper.api.backend.SilverBackend.%24anonfun%24submit%241%28SilverBackend.scala%3A164%29%0A++at+viper.api.backend.SilverBackend.%24anonfun%24submit%241%24adapted%28SilverBackend.scala%3A164%29%0A++at+scala.collection.immutable.List.foreach%28List.scala%3A333%29%0A++at+viper.api.backend.SilverBackend.submit%28SilverBackend.scala%3A164%29%0A++...%0A++at+vct.main.Main%24.main%28Main.scala%3A50%29%0A++at+vct.main.Main.main%28Main.scala%29%0A%60%60%60%0A%0A%23+Version+Information%0A*+VerCors+version+%602.3.0%60%0A*+At+commit+4bd0a112c13d35dc64e9b357e063afa5af678e6b+from+branch+%60HEAD%60+%28changes%3Dfalse%29%0A%0A%23+Arguments%0A%60%2Ftmp%2F2aeef290-57e7-11f0-8239-9fbcc99b708f%2Ftest.java%60+%60--progress%60%0A%0A%23+File+Inputs%0A%3Cdetails%3E%0A%3Csummary%3Etest.java%3C%2Fsummary%3E%0A%0A%60%60%60java%0A%2F%2F+-*-+tab-width%3A2+%3B+indent-tabs-mode%3Anil+-*-%0A%2F%2F%3A%3A+cases+ZeroArrayIC%0A%2F%2F%3A%3A+tools+silicon%0A%2F%2F%3A%3A+verdict+Pass%0A%0Apublic+class+ZeroArray+%7B%0A++++%2F*%40+%0A++++++++context_everywhere+ar+%21%3D+null%3B%0A++++++++requires+Perm%28ar%5B*%5D%2C1%29%3B%0A++++++++ensures+Perm%28ar%5B*%5D%2C1%29+**+%28%5Cforall+int+k+%3B+0+%3C%3D+k+%26%26+k+%3C+ar.length+%3B+ar%5Bk%5D%3D%3D0+%29+%3B%0A+++++*%2F%0A++++public+void+zero_array%28int+ar%5B%5D%29%7B%0A++++++++int+N%3Dar.length%3B%0A++++++++for%28int+i%3D0%3Bi%3CN%3Bi%2B%2B%29%0A++++++++%2F*%40++requires+Perm%28ar%5Bi%5D%2C1%29%3B%0A+++++++++++++ensures++Perm%28ar%5Bi%5D%2C1%29+**+ar%5Bi%5D%3D%3D0%3B%0A++++++++%40*%2F%0A%09%09++++++++%7B%0A++++++++++++ar%5Bi*2%5D%3D0%3B%0A++++++++%7D%0A%09%09for%28int+i%3D1%3Bi%3CN%3Bi%2B%2B%29%0A++++++++%2F*%40++requires+Perm%28ar%5Bi%5D%2C1%29%3B%0A+++++++++++++ensures++Perm%28ar%5Bi%5D%2C1%29+**+ar%5Bi%5D%3D%3D0%3B%0A++++++++%40*%2F%0A%0A++++++++%7B%0A++++++++++++ar%5Bi*2%5D%3D0%3B%0A++++++++%7D%0A++++%7D%0A%7D%0A%0A%0A%60%60%60%0A%3C%2Fdetails%3E%0A%0A%0A%23+Full+Log%0A%3Cdetails%3E%0A%0A%60%60%60%0A08%3A24%3A43.190+%5BINFO%5D+Start%3A+VerCors+%28at+08%3A24%3A43%29%0A08%3A25%3A30.021+%5BINFO%5D+Done%3A+VerCors+%28at+08%3A25%3A30%2C+duration%3A+00%3A00%3A46%29%0A08%3A25%3A30.078+%5BWARN%5D+Non-daemon+thread+UnknownThreadGroup.%5BVerCors%5D+Block+layout+updates+%28%2314%29+is+still+running%0A08%3A25%3A30.079+%5BWARN%5D+Due+to+this+VerCors+will+not+exit+and+sit+idly+until+that+thread+is+done.+You+may+want+to+stop+the+program+manually.%0A08%3A25%3A30.099+%5BERROR%5D+vct.col.origin.BlameUnreachable%3A+An+error+condition+was+reached%2C+which+should+be+statically+unreachable.+Assigning+to+a+subscript+should+trigger+an+error+on+the+assignment%2C+and+not+on+the+subscript..+Inner+failure%3A%0A+%3E+%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%0A+%3E+At+test.java%0A+%3E+--------------------------------------%0A+%3E++++25++%0A+%3E++++26++++++++++%7B%0A+%3E+++++++++++++++++++%5B-------%0A+%3E++++27++++++++++++++ar%5Bi*2%5D%3D0%3B%0A+%3E++++++++++++++++++++-------%5D%0A+%3E++++28++++++++++%7D%0A+%3E++++29++++++%7D%0A+%3E+--------------------------------------%0A+%3E+Index+may+be+negative%2C+or+exceed+the+length+of+the+array.+%28https%3A%2F%2Futwente.nl%2Fvercors%23arrayBounds%29%0A+%3E+%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%0A%09at+vct.col.origin.PanicBlame.blame%28Blame.scala%3A1460%29%0A%09at+vct.col.rewrite.adt.ImportArray%24ArrayBoundsPreconditionFailed.blame%28ImportArray.scala%3A27%29%0A%09at+vct.col.rewrite.adt.ImportArray%24ArrayBoundsPreconditionFailed.blame%28ImportArray.scala%3A22%29%0A%09at+vct.col.origin.NoContext.blame%28Blame.scala%3A1592%29%0A%09at+vct.col.origin.NoContext.blame%28Blame.scala%3A1588%29%0A%09at+viper.api.backend.SilverBackend.processError%28SilverBackend.scala%3A207%29%0A%09at+viper.api.backend.SilverBackend.processError%24%28SilverBackend.scala%3A172%29%0A%09at+viper.api.backend.silicon.Silicon.processError%28Silicon.scala%3A34%29%0A%09at+viper.api.backend.SilverBackend.%24anonfun%24submit%241%28SilverBackend.scala%3A164%29%0A%09at+viper.api.backend.SilverBackend.%24anonfun%24submit%241%24adapted%28SilverBackend.scala%3A164%29%0A%09at+scala.collection.immutable.List.foreach%28List.scala%3A333%29%0A%09at+viper.api.backend.SilverBackend.submit%28SilverBackend.scala%3A164%29%0A%09at+viper.api.backend.SilverBackend.submit%24%28SilverBackend.scala%3A143%29%0A%09at+viper.api.backend.silicon.Silicon.submit%28Silicon.scala%3A34%29%0A%09at+vct.main.stages.SilverBackend.verify%28Backend.scala%3A201%29%0A%09at+vct.main.stages.Backend.%24anonfun%24run%246%28Backend.scala%3A166%29%0A%09at+vct.main.stages.Backend.%24anonfun%24cachedDefinitelyVerifiesOrElseUpdate%241%28Backend.scala%3A106%29%0A%09at+scala.Option.getOrElse%28Option.scala%3A201%29%0A%09at+vct.main.stages.Backend.cachedDefinitelyVerifiesOrElseUpdate%28Backend.scala%3A104%29%0A%09at+vct.main.stages.Backend.cachedDefinitelyVerifiesOrElseUpdate%24%28Backend.scala%3A100%29%0A%09at+vct.main.stages.SilverBackend.cachedDefinitelyVerifiesOrElseUpdate%28Backend.scala%3A184%29%0A%09at+vct.main.stages.Backend.%24anonfun%24run%245%28Backend.scala%3A166%29%0A%09at+vct.main.stages.Backend.%24anonfun%24run%245%24adapted%28Backend.scala%3A163%29%0A%09at+hre.progress.Progress%24.%24anonfun%24foreach%242%28Progress.scala%3A25%29%0A%09at+scala.runtime.java8.JFunction0%24mcV%24sp.apply%28JFunction0%24mcV%24sp.scala%3A18%29%0A%09at+hre.progress.task.AbstractTask.frame%28AbstractTask.scala%3A166%29%0A%09at+hre.progress.Progress%24.%24anonfun%24foreach%241%28Progress.scala%3A25%29%0A%09at+hre.progress.Progress%24.%24anonfun%24foreach%241%24adapted%28Progress.scala%3A24%29%0A%09at+scala.collection.IterableOnceOps.foreach%28IterableOnce.scala%3A576%29%0A%09at+scala.collection.IterableOnceOps.foreach%24%28IterableOnce.scala%3A574%29%0A%09at+scala.collection.AbstractIterator.foreach%28Iterator.scala%3A1300%29%0A%09at+hre.progress.Progress%24.foreach%28Progress.scala%3A24%29%0A%09at+vct.main.stages.Backend.%24anonfun%24run%241%28Backend.scala%3A163%29%0A%09at+hre.progress.task.NameSequenceTask.scope%28NameSequenceTask.scala%3A16%29%0A%09at+hre.progress.Progress%24.stages%28Progress.scala%3A47%29%0A
```
I am doing this herewith.

--- a/src/col/vct/col/origin/Blame.scala
+++ b/src/col/vct/col/origin/Blame.scala
@@ -1607,10 +1607,6 @@ object DerefAssignTarget
     extends PanicBlame(
       "Assigning to a field should trigger an error on the assignment, and not on the dereference."
     )
-object SubscriptAssignTarget
-    extends PanicBlame(
-      "Assigning to a subscript should trigger an error on the assignment, and not on the subscript."
-    )
 object DerefPerm
     extends PanicBlame(
       "Dereferencing a field in a permission should trigger an error on the permission, not on the dereference."

--- a/src/rewrite/vct/rewrite/ResolveExpressionSideEffects.scala
+++ b/src/rewrite/vct/rewrite/ResolveExpressionSideEffects.scala
@@ -11,7 +11,6 @@ import vct.col.origin.{
   LabelContext,
   Origin,
   PreferredName,
-  SubscriptAssignTarget,
 }
 import vct.col.ref.Ref
 import vct.col.rewrite.{
@@ -515,17 +514,17 @@ case class ResolveExpressionSideEffects[Pre <: Generation]()
           SilverDeref[Post](notInlined(obj), succ(f))(DerefAssignTarget)(
             target.o
           )
-        case ArraySubscript(arr, index) =>
-          ArraySubscript[Post](notInlined(arr), notInlined(index))(
-            SubscriptAssignTarget
-          )(target.o)
+        case sub @ ArraySubscript(arr, index) =>
+          ArraySubscript[Post](notInlined(arr), notInlined(index))(sub.blame)(
+            target.o
+          )
         case PointerSubscript(arr, index)
             if arr.t.isInstanceOf[TConstPointer[_]] =>
           throw DisallowedAssignmentTarget(target)
-        case PointerSubscript(arr, index) =>
-          PointerSubscript[Post](notInlined(arr), notInlined(index))(
-            SubscriptAssignTarget
-          )(target.o)
+        case sub @ PointerSubscript(arr, index) =>
+          PointerSubscript[Post](notInlined(arr), notInlined(index))(sub.blame)(
+            target.o
+          )
         case deref @ DerefPointer(ptr) =>
           DerefPointer[Post](notInlined(ptr))(deref.blame)(target.o)
         case VectorSubscript(_, _) => throw DisallowedAssignmentTarget(target)

